Bugfix: Configuration tasks should handle service restart

ClementJ35 opened the pull request:
# Bugfix: Configuration tasks should handle service restart

## 🗣 Description ##

Adding two tasks to stop the service concerned by a configuration modification, then restart it if it was started.

## 💭 Motivation and context ##

We need to apply configuration just after tweaking it. Handlers can't be used because of the freshclam check.

## 🧪 Testing ##

Simple test on running systems. It is simple enough to not need external tests (ansible.builtin.service).

## ✅ Pre-approval checklist ##

- [X] This PR has an informative and human-readable title.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [X] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [X] All relevant type-of-change labels have been added.
- [X] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [X] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [X] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.

ClementJ35 committed:
[configure] handle service restart if configuration is done

jsf9k commented on tasks/configure.yml:
```diff
@@ -1,4 +1,5 @@
 ---
+
```
Remove unnecessary blank line:
```suggestion
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -42,3 +43,15 @@
   ansible.builtin.file:
     mode: 0700
     path: "{{ task_conf_file }}"
+
+- name: Stop service {{ task_conf_service }}
+  ansible.builtin.service:
+    name: "{{ task_conf_service }}"
+    state: stopped
+  register: service_status
+
+- name: Restart service if it was started
+  ansible.builtin.service:
+    name: "{{ task_conf_service }}"
+    state: started
+  when: service_status.changed
```
This should be done in a handler instead.  You can probably just `notify` the handler in `tasks.main.yml` if and only if `configure.yml` actually modifies the configuration file in question.

ClementJ35 replied:
Indeed, it's cleaner. I was facing an issue to test services state. But I see now. I'll be back.

ClementJ35 replied:
resolved with [4d1c132](https://github.com/cisagov/ansible-role-clamav/pull/75/commits/4d1c1324f032a5b8451f456c1053382e55f4c9e0) commit, using handlers

jsf9k reviewed (changes requested):
Thanks for the contribution, @ClementJ35!

ClementJ35 committed:
Remove unnecessary blank line

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

ClementJ35 committed:
[handlers] Make use of handlers to restart services

We have to flush them early though, to pass signatures downloading
check if freshclam needs tweaks to operate, like using a private repo.

ClementJ35 committed:
[configure] passing notify on top of block to satisfy linting,

it breaks the alphabetic order mantra.

jsf9k commented on handlers/main.yml:
```diff
@@ -0,0 +1,22 @@
+---
+- name: Handle freshclam service
+  ansible.builtin.service:
+    name: "{{ freshclam_service_name }}"
+    state: stopped
+  notify: Start freschlam
+
+- name: Start freschlam
+  ansible.builtin.service:
+    name: "{{ freshclam_service_name }}"
+    state: started
+
+- name: Handle clamd service
+  ansible.builtin.service:
+    name: "{{ freshclam_service_name }}"
+    state: stopped
+  notify: Start clamd
+
+- name: Start clamd
+  ansible.builtin.service:
+    name: "{{ freshclam_service_name }}"
+    state: started
```
Why not just this?
```suggestion
- name: Restart freshclam service
  ansible.builtin.service:
    name: "{{ freshclam_service_name }}"
    state: restart

- name: Restart clamd service
  ansible.builtin.service:
    name: "{{ clamd_service_name }}"
    state: restart

ClementJ35 replied:
Because the restart state doesn't test that the service was started before or not. It start it whatever.

jsf9k commented on tasks/configure.yml:
```diff
@@ -1,44 +1,47 @@
-    - molecule-idempotence-notest
-  when: clamav_configuration_backup
+- name: Handle service configuration and notify restart if needed
+  notify: "Handle {{ task_conf_service }} service"
```
```suggestion
  notify: "Restart {{ task_conf_service }} service"
```

jsf9k commented on tasks/main.yml:
```diff
@@ -82,8 +83,14 @@
+    task_conf_service: clamd
   when: clamav_clamd_configuration
 
+# We need to flush handlers as configuration changes might be necessary
+# to pass the wait_for signatures test.
+- name: Flush possible pending handlers
+  ansible.builtin.meta: flush_handlers
+
```
I'd prefer this be moved to just before the "Wait for new signatures to be downloaded and installed by freshclam" task.

ClementJ35 replied:
This is not accurate. As I have accepted your modification to restart the service no matter what, we need to flush the handlers before the systemd task, which will set the attended status of services, and stop them if we need to.

If we flush after systemd, clamd will always be started if there is a configuration change.

ClementJ35 replied:
@jsf9k what do you think of my previous comment ?

jsf9k reviewed (changes requested):
Requesting some changes...

ClementJ35 committed:
[Handlers] Use basic restart instead of chaining start / stop handlers

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

ClementJ35 committed:
[Handlers] rename handlers notification to match restart

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

ClementJ35 committed:
[handlers] status must be restarted, not restart

ClementJ35 committed:
[handlers] fix clamav_service_name variable name

jsf9k commented on tasks/configure.yml:
```diff
@@ -1,44 +1,47 @@
-  tags:
-    - molecule-idempotence-notest
-  when: clamav_configuration_backup
+- name: Handle service configuration and notify restart if needed
```
Clean up language:
```suggestion
- name: Configure service and notify restart handler
```

jsf9k reviewed (changes requested):
One minor language change and then I think I am good.

jsf9k commented:
@ClementJ35 - Can you address my one new comment and then rebase on top of the latest?  Or if you prefer just address my one new comment and let me know that you want me to handle the rebasing.

ClementJ35 committed:
change task name to reflect handler name

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

ClementJ35 commented:
Hi @jsf9k , I did the changes and rebased.

dav3r reviewed (approved):
👍
